pkg/pillar/types/errortime.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package types
+
+import (
+	"reflect"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Common error with timestamp
+
+// ErrorAndTime is used by many EVE agents
+type ErrorAndTime struct {
+	Error     string
+	ErrorTime time.Time
+}
+
+// SetErrorNow uses the current time
+func (etPtr *ErrorAndTime) SetErrorNow(errStr string) {
+	etPtr.Error = errStr
+	etPtr.ErrorTime = time.Now()
+}
+
+// SetError is when time is specified
+func (etPtr *ErrorAndTime) SetError(errStr string, errorTime time.Time) {
+	etPtr.Error = errStr
+	etPtr.ErrorTime = errorTime
+}
+
+// ClearError removes it
+func (etPtr *ErrorAndTime) ClearError() {
+	etPtr.Error = ""
+	etPtr.ErrorTime = time.Time{}
+}
+
+// ErrorAndTimeWithSource has an additional field "ErrorSourceType"
+// which is used to selectively clear errors by calling IsErrorSource before
+// calling ClearErrorWithSource. See zedmanager and volumemgr for example use.
+type ErrorAndTimeWithSource struct {
+	ErrorSourceType interface{}
+	Error           string
```

Comment:
Why not use ErrorAndTime as an embedded field? You will not have to re-define all the methods then..

type ErrorAndTimeWithSource struct {
   ErrorAndTime
   ErrorSourceType interface{}
}

You will get SetError and SetErrorNow for free because of this embedded structure

Reply:
I want to keep them as separately as possible to reduce the risk for future confusion.
Note that I don't have a SetErrorNow() and the Clear function has a different name very intentionally to make the callers more readable.

Reply:
Hmm.. My guess is - When reading code - mentally - this is all "Error Information" - You tend to skip the details when following through the code.. 

Now that we split this into 2 separate structures - I think the risk of confusion is pretty much zero. We can also add comments to indicate which one to use when.

The way u defined it is very clear - ErrorWithSource has specific purpose - only when u have to store the specific source along with Error information.. Having ErrorAndTime as an embedded struct here is not going to cause confusion..

Similarly - When clearing Error information - even for ErrorAndTimeWithSource - For the user, it is all error information and related - hence "ClearError()" should be more than enough..

just my $0.02

Reply:
@kalyan-nidumolu if you look at the code pattern where ClearErrorWithSource is used, which is always together with IsErrorSource, it seems more clear to me than if that is a ClearError - what does that function (which is defined for another type) have to do with the source matching one would ask.

Reply:
> Hmm.. My guess is - When reading code - mentally - this is all "Error Information" - You tend to skip the details when following through the code..

FWIW that seems like a very dangerous approach, which is why specific concepts needs to have specific names and not generic-sounding ones.

Its all liquids - you tend to skip the details. Tea, beer, gasoline, pesticides. All liquides.